internal/api/errors.go

```diff
@@ -0,0 +1,299 @@
+		}
+
+	case t_api.ErrFailedToUpdatePromise:
+		apiError = APIError{
+			Code:    http.StatusInternalServerError,
+			Message: "Failed to update promise",
+			Status:  err.Code().String(),
+		}
+
+	case t_api.ErrAIONetworkFailure:
+		apiError = APIError{
+			Code:    http.StatusInternalServerError,
+			Message: "Network failure",
+			Status:  err.Code().String(),
+		}
+
+	case t_api.ErrAIOStoreFailure:
+		apiError = APIError{
+			Code:    http.StatusInternalServerError,
+			Message: "Store failure",
+			Status:  err.Code().String(),
+		}
+	default:
+		apiError = APIError{
+			Code: http.StatusInternalServerError,
+		}
+	}
+
+	apiError.Details = append(apiError.Details, ErrorDetail{
+		Type:    "ResonateError",
+		Message: err.Error(),
+		Domain:  "platform",
+		Metadata: map[string]string{
+			"url": fmt.Sprintf("https://docs.resonatehq.io/reference/error-codes#%s", apiError.Status),
+		},
+	})
+
+	return &APIErrorResponse{APIError: apiError}
+}
+
+func IsApplicationLevelError(status t_api.ResponseStatus) bool {
+	switch status {
+	case t_api.StatusOK, t_api.StatusCreated, t_api.StatusNoContent:
+		return false
+	default:
+		return true
+	}
+}
+
+// HandleApplicationLevelError handles application level errors and returns an APIErrorResponse.
+func HandleApplicationLevelError(status t_api.ResponseStatus) *APIErrorResponse {
+	var apiError APIError
+
+	switch status {
+
+	// 2xx
+
+	case t_api.StatusOK, t_api.StatusCreated, t_api.StatusNoContent:
+		return nil
+
+	// 4xx
+
+	case t_api.StatusPromiseAlreadyResolved:
+		apiError = APIError{
+			Code:    http.StatusForbidden,
+			Message: "The promise has already been resolved",
+			Status:  t_api.StatusPromiseAlreadyResolved.String(),
+		}
+
+	case t_api.StatusPromiseAlreadyRejected:
+		apiError = APIError{
+			Code:    http.StatusForbidden,
+			Message: "The promise has already been rejected",
+			Status:  t_api.StatusPromiseAlreadyRejected.String(),
+		}
+
+	case t_api.StatusPromiseAlreadyCanceled:
+		apiError = APIError{
+			Code:    http.StatusForbidden,
+			Message: "The promise has already been canceled",
+			Status:  t_api.StatusPromiseAlreadyCanceled.String(),
+		}
+
+	case t_api.StatusPromiseAlreadyTimedOut:
+		apiError = APIError{
+			Code:    http.StatusForbidden,
+			Message: "The promise has already timed out",
+			Status:  t_api.StatusPromiseAlreadyTimedOut.String(),
+		}
+
+	case t_api.StatusPromiseNotFound:
+		apiError = APIError{
+			Code:    http.StatusNotFound,
+			Message: "The specified promise was not found",
+			Status:  t_api.StatusPromiseNotFound.String(),
+		}
+
+	case t_api.StatusPromiseAlreadyExists:
+		apiError = APIError{
+			Code:    http.StatusConflict,
+			Message: "A promise with this identifier already exists",
+			Status:  t_api.StatusPromiseAlreadyExists.String(),
+		}
+
+	default:
+		panic(fmt.Sprintf("unknown status code %d", status))
+	}
+
+	apiError.Details = append(apiError.Details, ErrorDetail{
+		Type:    "ApplicationLevelError",
```

Comment:
I'm not sure "application level" is the best way to communicate with the client, any other ideas for what we could call this class of errors?